runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java

```diff
@@ -311,8 +315,26 @@ public DataflowPackage stageToFile(
       CompletionStage<StagingResult> stagingResult =
           computePackageAttributes(source, hash, dest, stagingPath)
               .thenComposeAsync(
-                  packageAttributes ->
-                      stagePackage(packageAttributes, retrySleeper, createOptions));
+                  packageAttributes -> {
+                    String destLocation = packageAttributes.getDestination().getLocation();
+                    String existingHash =
+                        distinctDestinations.putIfAbsent(destLocation, packageAttributes.getHash());
+                    if (existingHash == null) {
+                      return stagePackage(packageAttributes, retrySleeper, createOptions);
+                    } else {
+                      if (!existingHash.equals(packageAttributes.getHash())) {
+                        LOG.warn(
+                            "Upload of {} would overwrite {} with different content",
+                            packageAttributes.getSource(),
+                            destLocation);
```

Comment:
Should this throw an exception? IIUC, this would most likely mean that a user-specified file to upload would fail, right? (classpath JAR files with equal names should have equal contents as they usually include version in their name)

Wouldn't it be best to error out, and inform the use about that?

Reply:
I'm cool with either way.  I left it like this because in the past it wouldn't be an error, they'd just end up with possibly unexpected results.

Reply:
Two files should not have a same destination in the first place. They are deduplicated by postfixing hash when dependency information is created: https://github.com/apache/beam/blob/master/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java#L320

Reply:
Regarding my previous comment, I prefer failing the job instead of just printing out warnings when two files have a same staging destination.

Reply:
yeah good call, this can never be hit since the destination names will never be the same.  I'll just remove the dupe check completely.

Reply:
all set